SingularityService/src/main/java/com/hubspot/singularity/data/ApiCache.java

```diff
@@ -0,0 +1,77 @@
+package com.hubspot.singularity.data;
+
+import com.google.inject.Inject;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ApiCache<K, V> {
+  private static final Logger LOG = LoggerFactory.getLogger(ApiCache.class);
+
+  public final boolean isEnabled;
+  private final AtomicReference<Map<K, V>> zkValues;
+  private final Supplier<Map<K, V>> supplyMap;
+
+  @Inject
+  public ApiCache(
+    boolean isEnabled,
+    int cacheTtl,
+    Supplier<Map<K, V>> supplyMap,
+    ScheduledExecutorService executor
+  ) {
+    this.isEnabled = isEnabled;
+    this.supplyMap = supplyMap;
+    this.zkValues = new AtomicReference<>(new HashMap<>());
+
+    if (this.isEnabled) {
+      executor.scheduleAtFixedRate(this::reloadZkValues, 2, cacheTtl, TimeUnit.SECONDS);
```

Comment:
Put in a 2 second delay because the reloading thread kept starting before the CuratorFrameworkImpl started

Reply:
would it make more sense to try and do the first call to these inside some of the Managed classses that run during startup. Then we can:
- Make sure that the first fetch of this runs after curator has started
- Avoid writing zk calls in a constructor context vs some other method on an already instantiated class
- Have the cached filled before the service actually starts listening for requests
Thinking something like a start() method that we can call somewhere in SingularityLifecycleManaged